### PR TITLE
Add GCP OSD cluster installation

### PR DESCRIPTION
**Pull request by rnetser**
Requires https://github.com/RedHatQE/openshift-cluster-management-python-wrapper/pull/146

### CODE REVIEW

File: openshift_cli_installer/cli.py [hunk 1]
```diff
@@ -270,6 +284,24 @@ def main(**kwargs):
         for _region in _regions_to_verify:
             set_and_verify_aws_credentials(region_name=_region)
 
+    if gcp_osd_clusters:
```
Comment (myakove) [inline openshift_cli_installer/cli.py:287]: Extract to function

File: openshift_cli_installer/libs/managed_clusters/helpers.py [hunk 2]
```diff
@@ -9,20 +15,34 @@ def prepare_managed_clusters_data(
...
+
         _cluster["multi-az"] = _cluster.get("multi-az", False)
-        if _cluster["platform"] == HYPERSHIFT_STR:
+        if cluster_platform == HYPERSHIFT_STR:
             _cluster["hosted-cp"] = "true"
             _cluster["tags"] = "dns:external"
             _cluster["machine-cidr"] = _cluster.get("cidr", "10.0.0.0/16")
 
+        if cluster_platform == AWS_OSD_STR:
+            _cluster["aws-access-key-id"] = aws_access_key_id
```
Comment (myakove) [inline openshift_cli_installer/libs/managed_clusters/helpers.py:35]: HYPERSHIFT should also have `aws-access-key-id` etc

File: openshift_cli_installer/libs/managed_clusters/helpers.py [hunk 3]
```diff
@@ -9,20 +15,34 @@ def prepare_managed_clusters_data(
...
+            _cluster["aws-secret-access-key"] = aws_secret_access_key
+            _cluster["aws-account-id"] = aws_account_id
+
+        if cluster_platform == GCP_OSD_STR:
+            if not _gcp_service_account_file:
```
Comment (myakove) [inline openshift_cli_installer/libs/managed_clusters/helpers.py:40]: should be done once outside the for loop

File: openshift_cli_installer/libs/managed_clusters/helpers.py [hunk 4]
```diff
@@ -31,3 +51,8 @@ def prepare_managed_clusters_data(
             )
 
     return clusters
+
+
+def get_service_account_file(gcp_service_account_file):
```
Comment (myakove) [inline openshift_cli_installer/libs/managed_clusters/helpers.py:56]: should be renamed, you return dict

File: openshift_cli_installer/libs/managed_clusters/osd_clusters.py [hunk 5]
```diff
@@ -16,20 +21,32 @@ def osd_create_cluster(cluster_data):
...
+        }
+        if cluster_platform == AWS_OSD_STR:
+            provision_osd_kwargs.update(
+                {
+                    "aws_access_key_id": "aws-access-key-id",
+                    "aws_account_id": cluster_data["aws-account-id"],
+                    "aws_secret_access_key": cluster_data["aws-secret-access-key"],
+                }
+            )
+        if cluster_platform == GCP_OSD_STR:
```
Comment (myakove) [inline openshift_cli_installer/libs/managed_clusters/osd_clusters.py:45]: elif

File: openshift_cli_installer/utils/cli_utils.py [hunk 6]
```diff
@@ -50,11 +51,21 @@ def get_clusters_by_type(clusters):
     aws_osd_clusters = [
         _cluster for _cluster in clusters if _cluster["platform"] == AWS_OSD_STR
     ]
-    return aws_ipi_clusters, rosa_clusters, hypershift_clusters, aws_osd_clusters
+    gcp_osd_clusters = [
```
Comment (myakove) [inline openshift_cli_installer/utils/cli_utils.py:54]: we should return dict, the code can be done in for loop for all

File: openshift_cli_installer/utils/cli_utils.py [hunk 7]
```diff
@@ -50,11 +51,21 @@ def get_clusters_by_type(clusters):
...
+        rosa_clusters,
+        hypershift_clusters,
+        aws_osd_clusters,
+        gcp_osd_clusters,
+    )
 
 
 def is_platform_supported(clusters):
-    supported_platform = (AWS_STR, ROSA_STR, HYPERSHIFT_STR, AWS_OSD_STR)
+    supported_platform = (AWS_STR, ROSA_STR, HYPERSHIFT_STR, AWS_OSD_STR, GCP_OSD_STR)
```
Comment (myakove) [inline openshift_cli_installer/utils/cli_utils.py:68]: we should return the full list of clusters with unsupported platform

File: openshift_cli_installer/utils/cli_utils.py [hunk 8]
```diff
@@ -148,7 +159,7 @@ def create_openshift_cluster(
         return rosa_create_cluster(
             cluster_data=cluster_data,
         )
-    elif cluster_platform == AWS_OSD_STR:
+    elif cluster_platform in [AWS_OSD_STR, GCP_OSD_STR]:
```
Comment (myakove) [inline openshift_cli_installer/utils/cli_utils.py:162]: `[AWS_OSD_STR, GCP_OSD_STR]` make it tuple to be consistent.

File: openshift_cli_installer/utils/cli_utils.py [hunk 9]
```diff
@@ -162,7 +173,7 @@ def destroy_openshift_cluster(cluster_data):
     elif cluster_platform in (ROSA_STR, HYPERSHIFT_STR):
         return rosa_delete_cluster(cluster_data=cluster_data)
 
-    elif cluster_platform == AWS_OSD_STR:
+    elif cluster_platform in [AWS_OSD_STR, GCP_OSD_STR]:
```
Comment (myakove) [inline openshift_cli_installer/utils/cli_utils.py:176]: ditto

File: openshift_cli_installer/utils/cli_utils.py [hunk 10]
```diff
@@ -276,10 +292,10 @@ def assert_aws_ipi_user_input(
         assert_registry_config_file_exists(registry_config_file=registry_config_file)
 
 
-def assert_osd_user_input(
+def assert_aws_osd_user_input(
     clusters, aws_access_key_id, aws_secret_access_key, aws_account_id
 ):
-    if any([_cluster["platform"] == AWS_OSD_STR for _cluster in clusters]):
+    if any([_cluster["platform"] == GCP_OSD_STR for _cluster in clusters]):
```
Comment (myakove) [inline openshift_cli_installer/utils/cli_utils.py:298]: `GCP_OSD_STR` should be `AWS_OSD_STR`?

File: openshift_cli_installer/utils/cli_utils.py [hunk 11]
```diff
@@ -421,3 +439,14 @@ def run_create_or_destroy_clusters(clusters, create, action, parallel):
             processed_clusters.append(result.result())
 
     return processed_clusters
+
+
+def assert_gcp_osd_user_input(clusters, gcp_service_account_file):
```
Comment (myakove) [inline openshift_cli_installer/utils/cli_utils.py:444]: we already have GCP clusters list, this function should get it

File: openshift_cli_installer/utils/cli_utils.py [hunk 12]
```diff
@@ -421,3 +439,14 @@ def run_create_or_destroy_clusters(clusters, create, action, parallel):
...
+    if any([_cluster["platform"] == GCP_OSD_STR for _cluster in clusters]):
+        if not gcp_service_account_file or not os.path.exists(gcp_service_account_file):
```
Comment (myakove) [inline openshift_cli_installer/utils/cli_utils.py:446]: No need to check if file exists, click do it for us
This function can be removed

File: openshift_cli_installer/utils/cluster_versions.py [hunk 13]
```diff
@@ -74,7 +78,7 @@ def set_clusters_versions(clusters, base_available_versions):
                 )
                 raise click.Abort()
 
-        click.echo(f"{cluster_name}: Cluster version set to {cluster_data['version']}")
+        LOGGER.info(f"{cluster_name}: Cluster version set to {cluster_data['version']}")
```
Comment (myakove) [inline openshift_cli_installer/utils/cluster_versions.py:81]: Remove LOGGER, we need to do it in another PR for all

File: openshift_cli_installer/utils/cluster_versions.py [hunk 14]
```diff
@@ -86,7 +90,10 @@ def filter_versions(version, base_versions_dict, platform, stream):
     versions_dict[stream] = {version_key: {"versions": set(), "latest": ""}}
 
     for _source, versions in base_versions_dict.items():
-        if platform in (HYPERSHIFT_STR, ROSA_STR, AWS_OSD_STR) and stream != _source:
+        if (
+            platform in (HYPERSHIFT_STR, ROSA_STR, AWS_OSD_STR, GCP_OSD_STR)
```
Comment (myakove) [inline openshift_cli_installer/utils/cluster_versions.py:94]: `(HYPERSHIFT_STR, ROSA_STR, AWS_OSD_STR, GCP_OSD_STR)` use supported_platform global variable (create it)

File: openshift_cli_installer/utils/gcp.py [hunk 15]
```diff
@@ -0,0 +1,14 @@
+from google.cloud import compute_v1
+from google.oauth2 import service_account
+
+
+def get_gcp_regions(gcp_service_account_file):
```
Comment (myakove) [inline openshift_cli_installer/utils/gcp.py:5]: add TODO to move this to cloud_tools

**Review by myakove — CHANGES_REQUESTED** [review]
Add tests for GCP versions

File: openshift_cli_installer/cli.py [hunk 16]
```diff
@@ -180,6 +182,14 @@
     """,
     type=click.Path(exists=True),
 )
+@click.option(
+    "----gcp-service-account-file",
+    help="""
+\b
+Path to GCP service account json file.
+""",
+    type=click.Path(),
```
[the remark this answers is not in the document]
Comment (rnetser) [reply openshift_cli_installer/cli.py:191]: done

**Review by myakove — DISMISSED** [review]

**Review by myakove — DISMISSED** [review]

**Review by myakove — APPROVED** [review]
